apps/condo/domains/contact/schema/Contact.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Generated by `createschema contact.Contact 'property:Relationship:Property:SET_NULL; name:Text; phone:Text; unitName?:Text; email?:Text;'`
+ */
+
+const { Text, Relationship } = require('@keystonejs/fields')
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+const { SENDER_FIELD, DV_FIELD } = require('@condo/domains/common/schema/fields')
+const { ORGANIZATION_OWNED_FIELD } = require('../../../schema/_common')
+const access = require('@condo/domains/contact/access/Contact')
+const { normalizePhone } = require('@condo/domains/common/utils/phone')
+
+
+const Contact = new GQLListSchema('Contact', {
+    schemaDoc: 'Contact information of a person. Currently it will be related to a ticket, but in the future, it will be associated with more things',
+    fields: {
+        dv: DV_FIELD,
+        sender: SENDER_FIELD,
+
+        organization: ORGANIZATION_OWNED_FIELD,
+
+        property: {
+            schemaDoc: 'Property, that is a subject of an issue, reported by this person in first ticket. Meaning of this field will be revised in the future',
+            type: Relationship,
+            ref: 'Property',
+            isRequired: false,
+            kmigratorOptions: { null: true, on_delete: 'models.SET_NULL' },
+        },
+
+        unitName: {
+            schemaDoc: 'Property unit, that is a subject of an issue, reported by this person in first ticket. Meaning of this field will be revised in the future',
+            type: Text,
+            isRequired: false,
+        },
+
+        email: {
+            schemaDoc: 'Normalized contact email of this person',
+            type: Text,
+            isRequired: false,
+            hooks: {
+                resolveInput: async ({ resolvedData }) => {
+                    return resolvedData['email'] && resolvedData['email'].toLowerCase()
+                },
+            },
+        },
+
+        phone: {
+            schemaDoc: 'Normalized contact phone of this person in E.164 format without spaces',
+            type: Text,
+            isRequired: true,
+            hooks: {
+                resolveInput: async ({ resolvedData }) => {
+                    return normalizePhone(resolvedData['phone'])
+                },
+            },
+        },
+
+        name: {
+            schemaDoc: 'Name or full name of this person',
+            type: Text,
+            isRequired: true,
+        },
+
+    },
+    plugins: [uuided(), versioned(), tracked(), softDeleted(), historical()],
+    access: {
```

Comment:
hook to check property + phone + unitName unique like in register new user

Reply:
I have implemented it on a database level.

Compare two implementations:

1. PostgreSQL — 1 line:

```
ALTER TABLE "Contact" ADD CONSTRAINT "Contact_uniq" UNIQUE ("property", "unitName", "name", "phone");
```

2. Keystone:

```
hooks: {
        validateInput: async ({ resolvedData, operation, existingItem, addValidationError, context }) => {
            const { property, unitName, name, phone } = resolvedData
            const [contact] = await ContactAPI.getAll(context, {
                property: { id: property },
                unitName,
                name,
                phone,
            })
            if (operation === 'create') {
                if (contact) {
                    return addValidationError('Cannot create contact, because another contact with the same provided set of "property", "unitName", "name", "phone"')
                }
            } else if (operation === 'update') {
                if (contact && contact.id !== existingItem.id) {
                    return addValidationError('Cannot update contact, because another contact already exists with the same provided set of "property", "unitName", "name", "phone"')
                }
            }
        },
    },
```

Replaced in `9bf20046` to Keystone solution for consistency, but, It is a sign of code bloating problems in the future.
As for me, a database is a full "citizen" in code base.